File: Kanishk-Tyagi/PrefixSum/matrixBlockSum.cpp
class Solution {
public:
    vector<vector<int>> matrixBlockSum(vector<vector<int>>& mat, int k) {
        int n = mat.size();
        int m = mat[0].size();

        vector<vector<int>> prefix(n + 1, vector<int>(m + 1, 0));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                prefix[i + 1][j + 1] = mat[i][j] 
                                     + prefix[i][j + 1] 
                                     + prefix[i + 1][j] 
                                     - prefix[i][j];
            }
        }

        vector<vector<int>> result(n, vector<int>(m, 0));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int tr = max(0, i - k);
                int br = min(n - 1, i + k);
                int lc = max(0, j - k);
                int rc = min(m - 1, j + k);

                tr++;
                br++;
                lc++;
                rc++;

                result[i][j] = prefix[br][rc]
                             - prefix[tr - 1][rc]
                             - prefix[br][lc - 1]
                             + prefix[tr - 1][lc - 1];
            }
        }

        return result;
    }
};

File: [Manya]/Leetcode/Prefix/matrix_block_sum.java
import java.util.Scanner;
import java.util.Arrays;

class Solution {
    public int[][] matrixBlockSum(int[][] mat, int k) {
        int m = mat.length;
        int n = mat[0].length;
        int[][] prefixSum = new int[m + 1][n + 1];
        
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                prefixSum[i][j] = mat[i - 1][j - 1] + prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1];
            }
        }
        
        int[][] result = new int[m][n];
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int r1 = Math.max(0, i - k);
                int r2 = Math.min(m - 1, i + k);
                int c1 = Math.max(0, j - k);
                int c2 = Math.min(n - 1, j + k);
                
                result[i][j] = prefixSum[r2 + 1][c2 + 1]
                             - prefixSum[r1][c2 + 1]
                             - prefixSum[r2 + 1][c1]
                             + prefixSum[r1][c1];
            }
        }
        
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of rows: ");
        int m = scanner.nextInt();
        System.out.print("Enter the number of columns: ");
        int n = scanner.nextInt();
        
        int[][] mat = new int[m][n];
        System.out.println("Enter the matrix elements:");
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = scanner.nextInt();
            }
        }
        
        System.out.print("Enter the value of k: ");
        int k = scanner.nextInt();
        
        Solution solution = new Solution();
        int[][] result = solution.matrixBlockSum(mat, k);
        
        System.out.println("Resultant matrix:");
        for (int[] row : result) {
            System.out.println(Arrays.toString(row));
        }
        
        scanner.close();
    }
}
